packages/fxa-auth-server/lib/payments/customer-locations.ts

```diff
@@ -0,0 +1,173 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { Logger } from 'mozlog';
+
+// Region codes from: https://en.wikipedia.org/wiki/ISO_3166-2:US and https://en.wikipedia.org/wiki/ISO_3166-2:CA
+import STATES_LONG_NAME_TO_SHORT_NAME_MAP from './states-long-name-to-short-name-map.json';
+import { StripeHelper } from './stripe';
+
+type PayPalUserLocationResult = {
+  uid: string;
+  state: string;
+  country: string;
+  count: number;
+};
+
+// The countries we need region data for
+const COUNTRIES_LONG_NAME_TO_SHORT_NAME_MAP = {
+  // The long name is used in the BigQuery metrics logs; the short name is used
+  // in the Stripe customer billing address
+  'United States': 'US',
+  Canada: 'CA',
+};
+
+function sleep(delayMs: number) {
+  return new Promise((resolve) => setTimeout(resolve, delayMs));
+}
+
+export class CustomerLocations {
+  private log: Logger;
+  private stripeHelper: StripeHelper;
+
+  constructor(log: Logger, stripeHelper: StripeHelper) {
+    this.log = log;
+    this.stripeHelper = stripeHelper;
+  }
+
+  // countryCode is an ISO 3166-2 code e.g. 'US' or 'CA'.
+  isCountryNeeded(countryCode: string) {
+    return Object.values(COUNTRIES_LONG_NAME_TO_SHORT_NAME_MAP).includes(
+      countryCode
+    );
+  }
+
+  /**
+   * For PayPal customers, their payment method country is stored in the
+   * customer's billing address. This will return an ISO 3166-2 country
+   *  code. 'US' and 'CA' are the only ones we're interested in.
+   */
+  async getPayPalCustomerCountry(uid: string): Promise<string | null> {
+    // 1. Get the Stripe customer object from Firestore else Stripe
+    const customer = await this.stripeHelper.fetchCustomer(uid);
+    if (!customer) {
+      this.log.debug('customerLocations.getPayPalCustomerCountry', {
+        message: 'No customer found in Stripe for the user.',
+        uid,
+      });
+      return null;
+    }
+    // 2. Get the Stripe customer billing address; should be `address.country`
+    // We already know this user is or was a PayPal subscriber, so we
```

Comment:
We do?  We rely on the calling code for that?

Reply:
Yeah for existing PayPal customers, the list of uids is coming from a [SQL query](https://mozilla-hub.atlassian.net/browse/FXA-4218?focusedCommentId=494580) off the `paypalCustomers` table in our FxA DB. More will be revealed tomorrow when we discuss #11070 .